### PR TITLE
[SPARK-17926][SQL][STREAMING] Added json for statuses

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

StreamingQueryStatus exposed through StreamingQueryListener often needs to be recorded (similar to SparkListener events). This PR adds `.json` and `.prettyJson` to `StreamingQueryStatus`, `SourceStatus` and `SinkStatus`.
## How was this patch tested?

New unit tests

### CODE REVIEW

**Comment by tdas** [comment]
@marmbrus

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/SourceStatus.scala [hunk 1]
```diff
@@ -47,8 +53,22 @@ class SourceStatus private(
...
   override def toString: String =
-    "SourceStatus:" + indent(prettyString)
+    "Status of source " + indent(prettyString).trim
+
+  private[sql] def jsonValue: JValue = {
+    ("description" -> JString(description)) ~
+    ("offsetDesc" -> JString(offsetDesc)) ~
+    ("inputRate" -> JDouble(inputRate)) ~
+    ("processingRate" -> JDouble(processingRate)) ~
+    ("triggerDetails" -> JsonProtocol.mapToJson(triggerDetails.asScala))
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/SourceStatus.scala]: Will triggerDetails have lots of stuff?

**Comment by yhuai** [comment]
lgtm1

**Comment by tdas** [comment]
Maybe 10-20, not much more. I dont think we will be adding a lot of stuff.

**Comment by yhuai** [comment]
lgtm

**Comment by tdas** [comment]
Thanks very much @yhuai !

**Pull request closed without merging**